pyvortex/src/dataset.rs

```diff
@@ -102,27 +101,31 @@ impl PyDataset {
 
         let row_filter = row_filter.map(|x| RowFilter::new(x.borrow().unwrap().clone()));
 
-        let layout_reader = TOKIO_RUNTIME
-            .block_on(io::layout_reader(
-                self_.fname(),
-                projection,
-                batch_size,
-                row_filter,
-            ))
-            .map_err(PyVortexError::map_err)?;
-
-        let record_batch_reader: Box<dyn RecordBatchReader + Send> = Box::new(
-            VortexRecordBatchReader::new(layout_reader, &TOKIO_RUNTIME)
-                .map_err(PyVortexError::map_err)?,
-        );
-
-        record_batch_reader.into_pyarrow(self_.py())
+        match &self_.file_or_url {
+            FileOrUrl::File(file) => {
+                let reader = TOKIO_RUNTIME.block_on(File::open(Path::new(file)))?;
+                Self::reader_to_pyarrow_record_batch_reader(
+                    self_, reader, projection, batch_size, row_filter,
+                )
+            }
+            FileOrUrl::Url(url) => {
+                let reader = TOKIO_RUNTIME.block_on(ObjectStoreReadAt::try_new_from_url(url))?;
+                Self::reader_to_pyarrow_record_batch_reader(
+                    self_, reader, projection, batch_size, row_filter,
+                )
+            }
+        }
     }
 }
 
 #[pyfunction]
-pub fn dataset(fname: &Bound<'_, PyString>) -> PyResult<PyDataset> {
-    TOKIO_RUNTIME
-        .block_on(PyDataset::new(fname.to_str()?))
-        .map_err(PyVortexError::map_err)
+#[pyo3(signature = (*, file = None, url = None))]
+pub fn dataset(
+    file: Option<&Bound<PyString>>,
```

Comment:
Having two mutually exclusive arguments is a bit odd in Python. Either it should be two functions, or a single argument that is switched on internally, e.g. `try { url = urlparse(url) } except { file = url }`

Reply:
@gatesn I switch to the separated to separate functions and merged main; ready for another look.